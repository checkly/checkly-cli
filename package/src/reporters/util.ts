import * as chalk from 'chalk'
import * as indentString from 'indent-string'
import { DateTime } from 'luxon'
import * as path from 'path'
import { Assertion } from '../constructs/api-check'

export enum CheckStatus {
  PENDING = 'pending',
  FAILED = 'failed',
  SUCCESSFUL = 'successful',
}

export function formatCheckTitle (status: CheckStatus, check: any) {
  const source = path.basename(check.sourceFile)
  let durationString = ''
  if (check.startedAt && check.stoppedAt) {
    const duration = DateTime.fromISO(check.stoppedAt)
      .diff(DateTime.fromISO(check.startedAt))
      .toHuman({ unitDisplay: 'narrow', listStyle: 'narrow' })
    durationString = `(${duration}) `
  }
  let title = `${status} ${durationString}> ${check.checkType.toLowerCase()} > ${source}`
  if (source !== check.name) {
    // For automatically generated Browser checks (.spec.js files), the source matches the name.
    // In this case, it looks nicer to only print the name once.
    title += `> ${check.name}`
  }
  if (status === CheckStatus.FAILED) {
    return chalk.bold.red(title)
  } else if (status === CheckStatus.PENDING) {
    return chalk.bold.magenta(title)
  } else {
    return chalk.bold.green(title)
  }
}

export function formatCheckResult (checkResult: any) {
  const result = []
  if (checkResult.checkType === 'API') {
    if (checkResult.checkRunData?.assertions?.length) {
      result.push(formatSectionTitle('Assertions'))
      result.push(formatAssertions(checkResult.checkRunData.assertions))
    }
    if (checkResult.checkRunData?.response) {
      result.push(formatSectionTitle('API Response'))
      result.push(formatApiResponse(checkResult.checkRunData.response))
    }
  }
  if (checkResult.logs?.length) {
    result.push(formatSectionTitle('Logs'))
    result.push(formatLogs(checkResult.logs))
  }
  return result.join('\n')
}

const assertionSources: any = {
  STATUS_CODE: 'status code',
  JSON_BODY: 'JSON body',
  HEADERS: 'headers',
  TEXT_BODY: 'text body',
  RESPONSE_TIME: 'response time',
}

const assertionComparisons: any = {
  EQUALS: 'equals',
  NOT_EQUALS: 'doesn\'t equal',
  HAS_KEY: 'has key',
  NOT_HAS_KEY: 'doesn\'t have key',
  HAS_VALUE: 'has value',
  NOT_HAS_VALUE: 'doesn\'t have value',
  IS_EMPTY: 'is empty',
  NOT_EMPTY: 'is not empty',
  GREATER_THAN: 'is greater than',
  LESS_THAN: 'is less than',
  CONTAINS: 'contains',
  NOT_CONTAINS: 'doesn\'t contain',
  IS_NULL: 'is null',
  NOT_NULL: 'is not null',
}

function formatAssertions (assertions: Array<Assertion&{ error: string, actual: any }>) {
  return assertions.map(({ source, property, comparison, target, regex, error, actual }) => {
    const assertionFailed = !!error
    const humanSource = assertionSources[source] || source
    const humanComparison = assertionComparisons[comparison] || comparison
    let actualString
    if (actual) {
      const { result: truncatedActual, lines: truncatedActualLines } = truncate(actual, {
        chars: 300,
        lines: 5,
        ending: chalk.magenta('\n...truncated...'),
      })

      // TODO: Clean this up somehow
      // If the string is multiple lines, we print it on a new line
      if (truncatedActualLines <= 1) {
        actualString = `Received: ${truncatedActual}.`
      } else {
        actualString = `Received:\n${indentString(truncatedActual, 4, { indent: ' ' })}`
      }
    }
    const message = [
      'Checking that the',
      humanSource,
      property ? `property "${property}"` : undefined,
      regex ? `regex "${regex}"` : undefined,
      humanComparison,
      `target "${target}"`,
      assertionFailed ? 'failed.' : 'succeeded.',
      actualString,
    ].filter(Boolean).join(' ')
    return assertionFailed ? chalk.red(message) : chalk.green(message)
  }).join('\n')
}

function formatApiResponse (response: any) {
  const { truncated, result: stringBody } = truncate(response.body, {
    chars: 20 * 100,
    lines: 20,
    ending: chalk.magenta('\n...truncated...'),
  })
  return [
    `Received ${response.status} ${response.statusText}`,
    response.body ? 'With body:' : undefined,
    stringBody, // TODO: Write to a local file if truncated.
  ].filter(Boolean).join('\n')
}

export function formatLogs (logs: Array<{ level: string, msg: string, time: number }>) {
  return logs.flatMap(({ level, msg, time }) => {
    const timestamp = DateTime.fromMillis(time).toLocaleString(DateTime.TIME_WITH_SECONDS)
    let format = chalk.dim
    if (level === 'WARN') {
      format = chalk.dim.yellow
    } else if (level === 'ERROR') {
      format = chalk.dim.red
    }
    const [firstLine, ...remainingLines] = msg.split('\n')
    return [
      `${level}, ${timestamp}, ${format(firstLine)}`,
      ...remainingLines.map((line) => format(line)),
    ]
  }).join('\n')
}

function formatSectionTitle (title: string): string {
  const width = process.stdout.isTTY ? process.stdout.columns - 18 : 80
  return `──${chalk.bold(title)}${'─'.repeat(width - title.length)}`
}

export function truncate (val: any, opts: { chars?: number, lines?: number, ending?: string }) {
  let truncated = false
  let result = toString(val)
  if (opts.chars && val.length > opts.chars) {
    truncated = true
    result = result.substring(0, opts.chars)
  }
  const lines = result.split('\n')
  if (opts.lines && lines.length > opts.lines) {
    truncated = true
    result = lines.slice(0, opts.lines).join('\n')
  }
  return {
    truncated,
    result: truncated && opts.ending ? result + opts.ending : result,
    lines: opts.lines ? Math.min(opts.lines, lines.length) : lines.length,
  }
}

export function toString (val: any): string {
  if (typeof val === 'object') {
    return JSON.stringify(val, null, 2)
  } else {
    return val.toString()
  }
}
