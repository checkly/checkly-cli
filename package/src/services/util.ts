import * as path from 'path'
import * as fs from 'fs/promises'
import { Service } from 'ts-node'

// TODO: Remove this in favor of glob? It's unused.
export async function walkDirectory (
  directory: string,
  ignoreDirectories: Set<string>,
  callback: (filepath: string) => Promise<void>,
): Promise<void> {
  const files = await fs.readdir(directory)
  for (const file of files.sort()) {
    const filepath = path.join(directory, file)
    const stats = await fs.stat(filepath)
    if (stats.isDirectory() && !ignoreDirectories.has(file)) {
      await walkDirectory(filepath, ignoreDirectories, callback)
    } else {
      await callback(filepath)
    }
  }
}

export async function loadJsFile (filepath: string): Promise<any> {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  let exported = require(filepath)
  if (exported instanceof Function) {
    exported = await exported()
  }
  return exported
}

export async function loadTsFile (filepath: string): Promise<any> {
  const tsCompiler = await getTsCompiler()
  tsCompiler.enabled(true)
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  let { default: exported } = require(filepath)
  if (exported instanceof Function) {
    exported = await exported()
  }
  tsCompiler.enabled(false) // Re-disable the TS compiler
  return exported
}

// To avoid a dependency on typescript for users with no TS checks, we need to dynamically import ts-node
let tsCompiler: Service
async function getTsCompiler (): Promise<Service> {
  if (tsCompiler) return tsCompiler
  try {
    const tsNode = await import('ts-node')
    tsCompiler = tsNode.register({
      compilerOptions: {
        module: 'CommonJS',
      },
    })
  } catch (err: any) {
    if (err.code === 'ERR_MODULE_NOT_FOUND' || err.code === 'MODULE_NOT_FOUND') {
      throw new Error('Please install ts-node and typescript to use TypeScript configuration files')
    }
    throw err
  }
  return tsCompiler
}
