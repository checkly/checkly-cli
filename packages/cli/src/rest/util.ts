import zlib from 'node:zlib'

import { AxiosRequestHeaders } from 'axios'
import { JsonStreamStringify } from 'json-stream-stringify'

export function compressJSONPayload (data: any, headers: AxiosRequestHeaders) {
  headers['Content-Type'] = 'application/json'
  headers['Content-Encoding'] = 'gzip'

  const zipper = zlib.createGzip()
  const streamer = new JsonStreamStringify(data)

  return streamer.pipe(zipper)
}
