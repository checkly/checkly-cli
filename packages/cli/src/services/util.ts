import * as path from 'path'
import * as fs from 'fs/promises'
import * as fsSync from 'fs'
import { Service } from 'ts-node'
import * as gitRepoInfo from 'git-repo-info'

export interface GitInformation {
  commitId: string
  branchName?: string | null
  commitOwner?: string | null
  commitMessage?: string | null
}

// TODO: Remove this in favor of glob? It's unused.
export async function walkDirectory (
  directory: string,
  ignoreDirectories: Set<string>,
  callback: (filepath: string) => Promise<void>,
): Promise<void> {
  const files = await fs.readdir(directory)
  for (const file of files.sort()) {
    const filepath = path.join(directory, file)
    const stats = await fs.stat(filepath)
    if (stats.isDirectory() && !ignoreDirectories.has(file)) {
      await walkDirectory(filepath, ignoreDirectories, callback)
    } else {
      await callback(filepath)
    }
  }
}

export async function loadJsFile (filepath: string): Promise<any> {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    let exported = require(filepath)
    if (exported instanceof Function) {
      exported = await exported()
    }
    return exported
  } catch (err: any) {
    throw new Error(`Error loading file ${filepath}\n${err.stack}`)
  }
}

export async function loadTsFile (filepath: string): Promise<any> {
  try {
    const tsCompiler = await getTsCompiler()
    tsCompiler.enabled(true)
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    let { default: exported } = require(filepath)
    if (exported instanceof Function) {
      exported = await exported()
    }
    tsCompiler.enabled(false) // Re-disable the TS compiler
    return exported
  } catch (err: any) {
    throw new Error(`Error loading file ${filepath}\n${err.stack}`)
  }
}

// To avoid a dependency on typescript for users with no TS checks, we need to dynamically import ts-node
let tsCompiler: Service
async function getTsCompiler (): Promise<Service> {
  if (tsCompiler) return tsCompiler
  try {
    const tsNode = await import('ts-node')
    tsCompiler = tsNode.register({
      compilerOptions: {
        module: 'CommonJS',
      },
    })
  } catch (err: any) {
    if (err.code === 'ERR_MODULE_NOT_FOUND' || err.code === 'MODULE_NOT_FOUND') {
      throw new Error('Please install ts-node and typescript to use TypeScript configuration files')
    }
    throw err
  }
  return tsCompiler
}

/**
 * @param relPath the path to be converted
 * @param separator this is for testing purposes only so we can reliably replace the separator on Linux / Darwin
 */
export function pathToPosix (relPath: string, separator?: string): string {
  // Windows uses \ rather than / as a path separator.
  // It's important that logical ID's are consistent across platforms, though.
  // Otherwise, checks will be deleted and recreated when `npx checkly deploy` is run on different machines.
  return path.normalize(relPath).split(separator ?? path.sep).join(path.posix.sep).replace(/^C:/, '')
}

export function splitConfigFilePath (configFile?: string): { configDirectory: string, configFilenames?: string[] } {
  if (configFile) {
    const cwd = path.resolve(path.dirname(configFile))
    return {
      configDirectory: cwd,
      configFilenames: [path.basename(configFile)],
    }
  }
  return {
    configDirectory: process.cwd(),
    configFilenames: undefined,
  }
}

export function isFileSync (path: string): boolean {
  // This helper is useful to test paths inside constructors which cannot be async.
  let result
  try {
    result = fsSync.existsSync(path)
  } catch (err: any) {
    throw new Error(`Error parsing the file path: ${path}`)
  }
  return result
}

export function getGitInformation (): GitInformation|null {
  const repositoryInfo = gitRepoInfo()

  if (!repositoryInfo.sha) {
    return null
  }

  return {
    commitId: repositoryInfo.sha,
    branchName: repositoryInfo.branch,
    commitOwner: repositoryInfo.committer,
    commitMessage: repositoryInfo.commitMessage,
  }
}
