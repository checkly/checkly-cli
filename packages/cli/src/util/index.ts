/**
 * Attempts to look up the value for the given secret and throws an error if
 * no value can be found.
 *
 * This basic utility is provided for convenience, but it is by no means best
 * practice and you are certainly not required to use it. We recommend
 * integrating your own secrets management instead.
 */
export function secret (name: string): string {
  const tryEnvs = new Set([
    `CHECKLY_SECRET_${name}`,
    `CHECKLY_SECRET_${name.toUpperCase()}`,
  ])

  for (const env of tryEnvs) {
    const envValue = process.env[env]
    if (envValue !== undefined) {
      return envValue
    }
  }

  const triedString = Array.from(tryEnvs.values()).join(', ')

  throw new Error(`Missing value for secret '${name}' (tried the following environment variables: ${triedString})`)
}
