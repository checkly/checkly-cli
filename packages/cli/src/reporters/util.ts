import * as chalk from 'chalk'
import * as indentString from 'indent-string'
import { DateTime } from 'luxon'
import * as logSymbols from 'log-symbols'

import { Assertion } from '../constructs/api-check'
import { getDefaults } from '../rest/api'

// eslint-disable-next-line no-restricted-syntax
export enum CheckStatus {
  PENDING,
  FAILED,
  SUCCESSFUL,
}

export function formatDuration (ms: number): string {
  if (ms < 1000) {
    return `${ms}ms`
  } else {
    return `${Math.ceil(ms / 1000)}s`
  }
}

export function formatCheckTitle (status: CheckStatus, check: any, opts: { includeSourceFile?: boolean } = {}) {
  let duration
  if (check.startedAt && check.stoppedAt) {
    const durationMs = DateTime.fromISO(check.stoppedAt)
      .diff(DateTime.fromISO(check.startedAt))
      .toMillis()
    duration = formatDuration(durationMs)
  }

  let statusString
  let format
  if (status === CheckStatus.SUCCESSFUL) {
    statusString = chalk.bold.green(logSymbols.success)
    format = chalk.bold
  } else if (status === CheckStatus.FAILED) {
    statusString = logSymbols.error
    format = chalk.bold.red
  } else {
    statusString = '-'
    format = chalk.bold.dim
  }

  return [
    format(statusString),
    opts.includeSourceFile ? format(`${check.sourceFile} >`) : undefined,
    format(check.name),
    duration ? `(${duration})` : undefined,
  ].filter(Boolean).join(' ')
}

export function formatCheckResult (checkResult: any) {
  const result = []
  if (checkResult.checkType === 'API') {
    // Order should follow the check lifecycle (response, then assertions)
    if (checkResult.checkRunData?.requestError) {
      result.push([
        formatSectionTitle('Request Error'),
        checkResult.checkRunData.requestError,
      ])
    } else {
      if (checkResult.checkRunData?.request) {
        result.push([
          formatSectionTitle('HTTP Request'),
          formatHttpRequest(checkResult.checkRunData.request),
        ])
      }
      if (checkResult.checkRunData?.response) {
        result.push([
          formatSectionTitle('HTTP Response'),
          formatHttpResponse(checkResult.checkRunData.response),
        ])
      }
      if (checkResult.checkRunData?.assertions?.length) {
        result.push([
          formatSectionTitle('Assertions'),
          formatAssertions(checkResult.checkRunData.assertions),
        ])
      }
      if (checkResult.logs?.setup.length) {
        result.push([
          formatSectionTitle('Setup Script Logs'),
          formatLogs(checkResult.logs.setup),
        ])
      }
      if (checkResult.logs?.teardown.length) {
        result.push([
          formatSectionTitle('Teardown Script Logs'),
          formatLogs(checkResult.logs.teardown),
        ])
      }
    }
  }
  if (checkResult.logs?.length) {
    result.push([
      formatSectionTitle('Logs'),
      formatLogs(checkResult.logs),
    ])
  }
  if (checkResult.runError) {
    result.push([
      formatSectionTitle('Execution Error'),
      formatRunError(checkResult.runError),
    ])
  }
  if (checkResult.scheduleError) {
    result.push([
      formatSectionTitle('Scheduling Error'),
      formatRunError(checkResult.scheduleError),
    ])
  }
  return result.map(([title, body]) => title + '\n' + body).join('\n\n')
}

const assertionSources: any = {
  STATUS_CODE: 'status code',
  JSON_BODY: 'JSON body',
  HEADERS: 'headers',
  TEXT_BODY: 'text body',
  RESPONSE_TIME: 'response time',
}

const assertionComparisons: any = {
  EQUALS: 'equals',
  NOT_EQUALS: 'doesn\'t equal',
  HAS_KEY: 'has key',
  NOT_HAS_KEY: 'doesn\'t have key',
  HAS_VALUE: 'has value',
  NOT_HAS_VALUE: 'doesn\'t have value',
  IS_EMPTY: 'is empty',
  NOT_EMPTY: 'is not empty',
  GREATER_THAN: 'is greater than',
  LESS_THAN: 'is less than',
  CONTAINS: 'contains',
  NOT_CONTAINS: 'doesn\'t contain',
  IS_NULL: 'is null',
  NOT_NULL: 'is not null',
}

function formatAssertions (assertions: Array<Assertion&{ error: string, actual: any }>) {
  return assertions.map(({ source, property, comparison, target, regex, error, actual }) => {
    const assertionFailed = !!error
    const humanSource = assertionSources[source] || source
    const humanComparison = assertionComparisons[comparison] || comparison
    let actualString
    if (actual) {
      const { result: truncatedActual, lines: truncatedActualLines } = truncate(actual, {
        chars: 300,
        lines: 5,
        ending: chalk.magenta('\n...truncated...'),
      })

      if (truncatedActualLines <= 1) {
        actualString = `Received: ${truncatedActual}.`
      } else {
        actualString = `Received:\n${indentString(truncatedActual, 4, { indent: ' ' })}`
      }
    }
    const message = [
      assertionFailed ? logSymbols.error : logSymbols.success,
      humanSource,
      property ? `property "${property}"` : undefined,
      regex ? `regex "${regex}"` : undefined,
      humanComparison,
      `target "${target}".`,
      actualString,
    ].filter(Boolean).join(' ')
    return assertionFailed ? chalk.red(message) : chalk.green(message)
  }).join('\n')
}

function formatHttpRequest (request: any) {
  const { truncated, result: stringBody } = truncate(request.data, {
    chars: 20 * 100,
    lines: 20,
    ending: chalk.magenta('\n...truncated...'),
  })
  const headersString = Object.entries(request.headers ?? []).map(([key, val]) => `${key}: ${val}`).join('\n')
  return [
    `${request.method} ${request.url}`,
    'Headers:',
    indentString(headersString, 2),
    request.data ? 'Body:' : undefined,
    indentString(stringBody, 2),
  ].filter(Boolean).join('\n')
}

function formatHttpResponse (response: any) {
  // TODO: Provide a user for a way to see the full response. For example, write it to a file.
  const { truncated, result: stringBody } = truncate(response.body, {
    chars: 20 * 100,
    lines: 20,
    ending: chalk.magenta('\n...truncated...'),
  })
  const headersString = Object.entries(response.headers ?? []).map(([key, val]) => `${key}: ${val}`).join('\n')
  return [
    `Status Code: ${response.status} ${response.statusText}`,
    'Headers:',
    indentString(headersString, 2),
    response.body ? 'Body:' : undefined,
    indentString(stringBody, 2),
  ].filter(Boolean).join('\n')
}

function formatLogs (logs: Array<{ level: string, msg: string, time: number }>) {
  return logs.flatMap(({ level, msg, time }) => {
    const timestamp = DateTime.fromMillis(time).toLocaleString(DateTime.TIME_24_WITH_SECONDS)
    let format = chalk.dim
    if (level === 'WARN') {
      format = chalk.dim.yellow
    } else if (level === 'ERROR') {
      format = chalk.dim.red
    }
    const [firstLine, ...remainingLines] = msg.split('\n')
    return [
      `${timestamp} ${level.padEnd(5, ' ')} ${format(firstLine)}`,
      ...remainingLines.map((line) => format(line)),
    ]
  }).join('\n')
}

function formatRunError (err: string | Error): string {
  if (typeof err === 'string') {
    return err
  } else {
    return err.message
  }
}

function formatSectionTitle (title: string): string {
  const width = process.stdout.isTTY ? Math.min(process.stdout.columns, 80) : 80
  return `──${chalk.bold(title)}${'─'.repeat(width - title.length)}`
}

function truncate (val: any, opts: { chars?: number, lines?: number, ending?: string }) {
  let truncated = false
  let result = toString(val)
  if (opts.chars && val.length > opts.chars) {
    truncated = true
    result = result.substring(0, opts.chars)
  }
  const lines = result.split('\n')
  if (opts.lines && lines.length > opts.lines) {
    truncated = true
    result = lines.slice(0, opts.lines).join('\n')
  }
  return {
    truncated,
    result: truncated && opts.ending ? result + opts.ending : result,
    lines: opts.lines ? Math.min(opts.lines, lines.length) : lines.length,
  }
}

function toString (val: any): string {
  if (typeof val === 'object') {
    return JSON.stringify(val, null, 2)
  } else {
    return val.toString()
  }
}

export function print (text: string) {
  process.stdout.write(text)
}

export function printLn (text: string, afterLnCount = 1, beforeLnCount = 0) {
  process.stdout.write(`${'\n'.repeat(beforeLnCount)}${text}${'\n'.repeat(afterLnCount)}`)
}

export function getTestSessionUrl (testSessionId: string): string {
  const { baseURL } = getDefaults()
  return `${baseURL.replace(/api/, 'app')}/test-sessions/${testSessionId}`
}

export function getTraceUrl (traceUrl: string): string {
  return `https://trace.playwright.dev/?trace=${encodeURIComponent(traceUrl)}`
}

export function printDeprecationWarning (text: string): void {
  console.log(chalk.yellow(`\n${logSymbols.warning} Deprecation warning: ${text} \n`))
}
