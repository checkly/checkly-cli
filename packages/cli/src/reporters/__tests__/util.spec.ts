import { formatCheckTitle, formatCheckResult, CheckStatus } from '../util'
import { simpleCheckFixture } from './fixtures/simple-check'
import { apiCheckResult } from './fixtures/api-check-result'
import { browserCheckResult } from './fixtures/browser-check-result'
import { Settings } from 'luxon'

function stripAnsi (input: string): string {
  return input.replace(
    // eslint-disable-next-line no-control-regex
    /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '')
}

beforeAll(() => {
  Settings.defaultZone = 'GMT'
})

describe('formatCheckTitle()', () => {
  it('should print a failed check title', () => {
    expect(stripAnsi(formatCheckTitle(CheckStatus.FAILED, simpleCheckFixture, { includeSourceFile: true })))
      .toMatchSnapshot('failed-check-title')
  })
  it('should print a passed check title', () => {
    expect(stripAnsi(formatCheckTitle(CheckStatus.SUCCESSFUL, simpleCheckFixture, { includeSourceFile: true })))
      .toMatchSnapshot('passed-check-title')
  })
  it('should print a pending check title', () => {
    expect(stripAnsi(formatCheckTitle(CheckStatus.PENDING, simpleCheckFixture, { includeSourceFile: true })))
      .toMatchSnapshot('pending-check-title')
  })
})

describe('formatCheckResult()', () => {
  describe('API Check result', () => {
    it('formats a basic API Check result ', () => {
      const basicApiCheckResult = { ...apiCheckResult }
      basicApiCheckResult.checkRunData.assertions = []
      basicApiCheckResult.logs.setup = []
      basicApiCheckResult.logs.teardown = []
      expect(stripAnsi(formatCheckResult(apiCheckResult)))
        .toMatchSnapshot('api-check-result-basic-format')
    })
    it('formats an API Check result with assertions', () => {
      const basicApiCheckResult = { ...apiCheckResult }
      basicApiCheckResult.logs.setup = []
      basicApiCheckResult.logs.teardown = []
      expect(stripAnsi(formatCheckResult(apiCheckResult)))
        .toMatchSnapshot('api-check-result-assertions-format')
    })
    it('formats an API Check result with setup & teardown logs', () => {
      const basicApiCheckResult = { ...apiCheckResult }
      basicApiCheckResult.checkRunData.assertions = []
      expect(stripAnsi(formatCheckResult(apiCheckResult)))
        .toMatchSnapshot('api-check-result-logs-format')
    })
    it('formats an API Check result with a scheduleError', () => {
      const basicApiCheckResult = { ...apiCheckResult }
      basicApiCheckResult.checkRunData.assertions = []
      basicApiCheckResult.logs.setup = []
      basicApiCheckResult.logs.teardown = []
      basicApiCheckResult.scheduleError = 'There was a scheduling error'
      expect(stripAnsi(formatCheckResult(basicApiCheckResult)))
        .toMatchSnapshot('api-check-result-schedule-error-format')
    })
  })
  describe('Browser Check result', () => {
    it('formats a basic Browser Check result ', () => {
      const basicBrowserCheckResult = { ...browserCheckResult }
      basicBrowserCheckResult.logs = []
      expect(stripAnsi(formatCheckResult(basicBrowserCheckResult)))
        .toMatchSnapshot('browser-check-result-basic-format')
    })
    it('formats a Browser Check result with logs', () => {
      expect(stripAnsi(formatCheckResult(browserCheckResult)))
        .toMatchSnapshot('browser-check-result-logs-format')
    })
    it('formats a Browser Check result with a scheduleError', () => {
      const basicBrowserCheckResult = { ...browserCheckResult }
      basicBrowserCheckResult.scheduleError = 'There was a scheduling error'
      expect(stripAnsi(formatCheckResult(basicBrowserCheckResult)))
        .toMatchSnapshot('browser-check-result-schedule-error-format')
    })
  })
})
