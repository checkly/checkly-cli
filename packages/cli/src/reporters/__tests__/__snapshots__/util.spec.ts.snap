// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatCheckResult() API Check result formats a basic API Check result : api-check-result-basic-format 1`] = `
"──HTTP Request────────────────────────────────────────────────────────────────────
POST https://httpbin.org/post
Headers:
  User-Agent: Checkly/1.0 (https://www.checklyhq.com)
  accept-encoding: gzip, deflate
  content-length: 10
Body:
  {"data":1}

──HTTP Response───────────────────────────────────────────────────────────────────
Status Code: 200 OK
Headers:
  date: Fri, 24 Mar 2023 10:45:42 GMT
  content-type: application/json
  content-length: 422
  connection: close
  server: gunicorn/19.9.0
  access-control-allow-origin: *
  access-control-allow-credentials: true
Body:
  {
    "args": {}, 
    "data": "{\\"data\\":1}", 
    "files": {}, 
    "form": {}, 
    "headers": {
      "Accept-Encoding": "gzip, deflate", 
      "Content-Length": "10", 
      "Host": "httpbin.org", 
      "User-Agent": "Checkly/1.0 (https://www.checklyhq.com)", 
      "X-Amzn-Trace-Id": "Root=1-641d7f55-0b99ca310b7322250bd32d00"
    }, 
    "json": {
      "data": 1
    }, 
    "origin": "3.251.180.115", 
    "url": "https://httpbin.org/post"
  }
"
`;

exports[`formatCheckResult() API Check result formats an API Check result with a scheduleError: api-check-result-schedule-error-format 1`] = `
"──HTTP Request────────────────────────────────────────────────────────────────────
POST https://httpbin.org/post
Headers:
  User-Agent: Checkly/1.0 (https://www.checklyhq.com)
  accept-encoding: gzip, deflate
  content-length: 10
Body:
  {"data":1}

──HTTP Response───────────────────────────────────────────────────────────────────
Status Code: 200 OK
Headers:
  date: Fri, 24 Mar 2023 10:45:42 GMT
  content-type: application/json
  content-length: 422
  connection: close
  server: gunicorn/19.9.0
  access-control-allow-origin: *
  access-control-allow-credentials: true
Body:
  {
    "args": {}, 
    "data": "{\\"data\\":1}", 
    "files": {}, 
    "form": {}, 
    "headers": {
      "Accept-Encoding": "gzip, deflate", 
      "Content-Length": "10", 
      "Host": "httpbin.org", 
      "User-Agent": "Checkly/1.0 (https://www.checklyhq.com)", 
      "X-Amzn-Trace-Id": "Root=1-641d7f55-0b99ca310b7322250bd32d00"
    }, 
    "json": {
      "data": 1
    }, 
    "origin": "3.251.180.115", 
    "url": "https://httpbin.org/post"
  }


──Scheduling Error────────────────────────────────────────────────────────────────
There was a scheduling error"
`;

exports[`formatCheckResult() API Check result formats an API Check result with assertions: api-check-result-assertions-format 1`] = `
"──HTTP Request────────────────────────────────────────────────────────────────────
POST https://httpbin.org/post
Headers:
  User-Agent: Checkly/1.0 (https://www.checklyhq.com)
  accept-encoding: gzip, deflate
  content-length: 10
Body:
  {"data":1}

──HTTP Response───────────────────────────────────────────────────────────────────
Status Code: 200 OK
Headers:
  date: Fri, 24 Mar 2023 10:45:42 GMT
  content-type: application/json
  content-length: 422
  connection: close
  server: gunicorn/19.9.0
  access-control-allow-origin: *
  access-control-allow-credentials: true
Body:
  {
    "args": {}, 
    "data": "{\\"data\\":1}", 
    "files": {}, 
    "form": {}, 
    "headers": {
      "Accept-Encoding": "gzip, deflate", 
      "Content-Length": "10", 
      "Host": "httpbin.org", 
      "User-Agent": "Checkly/1.0 (https://www.checklyhq.com)", 
      "X-Amzn-Trace-Id": "Root=1-641d7f55-0b99ca310b7322250bd32d00"
    }, 
    "json": {
      "data": 1
    }, 
    "origin": "3.251.180.115", 
    "url": "https://httpbin.org/post"
  }
"
`;

exports[`formatCheckResult() API Check result formats an API Check result with setup & teardown logs: api-check-result-logs-format 1`] = `
"──HTTP Request────────────────────────────────────────────────────────────────────
POST https://httpbin.org/post
Headers:
  User-Agent: Checkly/1.0 (https://www.checklyhq.com)
  accept-encoding: gzip, deflate
  content-length: 10
Body:
  {"data":1}

──HTTP Response───────────────────────────────────────────────────────────────────
Status Code: 200 OK
Headers:
  date: Fri, 24 Mar 2023 10:45:42 GMT
  content-type: application/json
  content-length: 422
  connection: close
  server: gunicorn/19.9.0
  access-control-allow-origin: *
  access-control-allow-credentials: true
Body:
  {
    "args": {}, 
    "data": "{\\"data\\":1}", 
    "files": {}, 
    "form": {}, 
    "headers": {
      "Accept-Encoding": "gzip, deflate", 
      "Content-Length": "10", 
      "Host": "httpbin.org", 
      "User-Agent": "Checkly/1.0 (https://www.checklyhq.com)", 
      "X-Amzn-Trace-Id": "Root=1-641d7f55-0b99ca310b7322250bd32d00"
    }, 
    "json": {
      "data": 1
    }, 
    "origin": "3.251.180.115", 
    "url": "https://httpbin.org/post"
  }
"
`;

exports[`formatCheckResult() Browser Check result formats a Browser Check result with a scheduleError: browser-check-result-schedule-error-format 1`] = `
"──Logs────────────────────────────────────────────────────────────────────────────
10:53:15 DEBUG Starting job
10:53:15 DEBUG Compiling environment variables
10:53:15 DEBUG Creating runtime using version 2022.10
10:53:15 DEBUG Running Playwright test script
10:53:16 INFO  Running 1 test using 1 worker
10:53:16 INFO  
10:53:18 INFO  [1/1] [chromium] › ../../var/task/src/2022-10/node_modules/vm2/lib/bridge.js:479:11 › visit page and take screenshot
10:53:22 INFO  
10:53:22 INFO  
10:53:22 INFO  1 passed (5s)
10:53:22 DEBUG Run finished
10:53:22 DEBUG Uploading log file

──Scheduling Error────────────────────────────────────────────────────────────────
There was a scheduling error"
`;

exports[`formatCheckResult() Browser Check result formats a Browser Check result with logs: browser-check-result-logs-format 1`] = `
"──Logs────────────────────────────────────────────────────────────────────────────
10:53:15 DEBUG Starting job
10:53:15 DEBUG Compiling environment variables
10:53:15 DEBUG Creating runtime using version 2022.10
10:53:15 DEBUG Running Playwright test script
10:53:16 INFO  Running 1 test using 1 worker
10:53:16 INFO  
10:53:18 INFO  [1/1] [chromium] › ../../var/task/src/2022-10/node_modules/vm2/lib/bridge.js:479:11 › visit page and take screenshot
10:53:22 INFO  
10:53:22 INFO  
10:53:22 INFO  1 passed (5s)
10:53:22 DEBUG Run finished
10:53:22 DEBUG Uploading log file"
`;

exports[`formatCheckResult() Browser Check result formats a basic Browser Check result : browser-check-result-basic-format 1`] = `""`;

exports[`formatCheckTitle() should print a failed check title: failed-check-title 1`] = `"✖ /test/test-file.check.ts > Test Check (11s)"`;

exports[`formatCheckTitle() should print a passed check title: passed-check-title 1`] = `"✔ /test/test-file.check.ts > Test Check (11s)"`;

exports[`formatCheckTitle() should print a pending check title: pending-check-title 1`] = `"- /test/test-file.check.ts > Test Check (11s)"`;
